README.md
# cupertino_stackview

[comment]: <> (Badges)
<a href="https://www.cosmossoftware.coffee">
   <img alt="Cosmos Software" src="https://img.shields.io/badge/Cosmos%20Software-Love%20Code-red" />
</a>
<a href="https://www.cosmossoftware.coffee">
   <img alt="Cosmos Software" src="https://img.shields.io/badge/Developer's-Choice-yellow" />
</a>

[![Pub](https://img.shields.io/pub/v/cupertino_stackview?color=g)](https://pub.dev/packages/cupertino_stackview)
[![License](https://img.shields.io/github/license/aliyigitbireroglu/flutter-cupertino-stackview?color=blue)](https://github.com/aliyigitbireroglu/flutter-cupertino-stackview/blob/master/LICENSE)

[comment]: <> (Introduction)
A very easy-to-use navigation tool/widget for having iOS 13 style stacks.

**It is highly recommended to read the documentation and run the example project on a real device to fully understand and inspect the full range of capabilities.**

[comment]: <> (ToC)
[Media](#media) | [Description](#description) | [Installation](#installation) | [How-to-Use](#howtouse)


[comment]: <> (Media)
<a name="media"></a>
## Media

Watch on **Youtube**:

[v1.0.0](https://youtu.be/XvwOScBMvEE)
<br><br>
<img src="https://www.cosmossoftware.coffee/Common/Portfolio/GIFs/FlutterCupertinoStackView.gif" height="450" max-height="450"/>
<br><br>


[comment]: <> (Description)
<a name="description"></a>
## Description
The "StackView" system, as I like to call it because I don't know the real name, is now the default navigation system for iOS 13. As a fan of this, I decided to implement a very easy-to-use version of it for Flutter as well.


[comment]: <> (How-to-Use)
<a name="howtouse"></a>
## How-to-Use
First, instantiate the cupertinoStackViewController that is provided by "misc.dart" with the required parameters:

```
cupertinoStackViewController = CupertinoStackViewController(
    navigatorState,
    {
      "Page I": (BuildContext context, dynamic parameters) {
        return PageI();
      },
      "Page II": (BuildContext context, dynamic parameters) {
        return PageII();
      },
      "Page III": (BuildContext context, dynamic parameters) {
        return PageIII();
      },
      "Page IV": (BuildContext context, dynamic parameters) {
        return PageIV();
      },
      ...,
    },
);
```

Then create a CupertinoStackView widget as shown in the example:

```
CupertinoStackView(
  true,             //_isPrimary 
  "Page I",         //_navigation
  Scaffold(...),    //_child
  Colors.black,     //_backgroundColor 
 {Key key,
  radius    : Radius.circular(10.0)})
```

**Further Explanations:**

*For a complete set of descriptions for all parameters and methods, see the [documentation](https://pub.dev/documentation/cupertino_stackview/latest/).*

* The [cupertinoStackViewController] from 'misc.dart' has to be initialised as soon as the application starts via the [void initialise(double height, double width)] method. Use this method with the [MediaQuery.of(context).size.height] and [MediaQuery.of(context).size.width] values.
* Use [CupertinoStackViewController]'s [Future back(BuildContext context, dynamic parameters) async] and [Future navigate(String targetNavigation, BuildContext context, dynamic parameters) async] methods for the corresponding actions. Do not call [Navigator] methods directly.


[comment]: <> (Notes)
## Notes
Any help, suggestion or criticism is appreciated! 

Cheers.

[comment]: <> (CosmosSoftware)
<br><br>
<img align="right" src="https://www.cosmossoftware.coffee/Common/Images/CosmosSoftwareIconTransparent.png" width="150" height="150"/>
<br><br>


cupertino_stackview/CHANGELOG.md
## [2.0.0] - 04.05.2021

* Migrated to Null Safety.

## [1.0.8] - 15.10.2020

* Minor changes.

## [1.0.7] - 11.10.2020

* Minor changes.

## [1.0.6] - 11.10.2020

* Regular maintenance.

## [1.0.5] - 01.06.2020

* Minor changes.

## [1.0.4] - 11.01.2020

* Minor changes.

## [1.0.3] - 05.01.2020

* Minor changes.

## [1.0.2] - 21.11.2019

* [Dismissable] is now implemented directly to the package. Use the [isDismissible] value to control it.

## [1.0.1] - 08.11.2019

* Minor changes.

## [1.0.0] - 06.11.2019

* Initial release.


cupertino_stackview/lib/misc.dart
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//  © Cosmos Software | Ali Yigit Bireroglu                                                                                                          /
//  All material used in the making of this code, project, program, application, software et cetera (the "Intellectual Property")                    /
//  belongs completely and solely to Ali Yigit Bireroglu. This includes but is not limited to the source code, the multimedia and                    /
//  other asset files.                                                                                                                               /
//  If you were granted this Intellectual Property for personal use, you are obligated to include this copyright text at all times.                  /
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

import 'package:cupertino_stackview/cupertino_stackview_controller.dart';

///See [CupertinoStackViewController].
late CupertinoStackViewController cupertinoStackViewController;

final double cupertinoStackViewScaleFraction = 2540.0 / 2688.0;
final double cupertinoStackViewFirstOffsetFraction = 180.0 / 2688.0;
final double cupertinoStackViewSecondOffsetFraction = 33.0 / 2688.0;
final double cupertinoStackViewSecondHeightFraction = 2510.0 / 2688.0;

enum CupertinoStackViewStatus {
  FRONT,
  BACK,
  DISAPPEAR,
}


cupertino_stackview/example/lib/main.dart
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//  © Cosmos Software | Ali Yigit Bireroglu                                                                                                          /
//  All material used in the making of this code, project, program, application, software et cetera (the "Intellectual Property")                    /
//  belongs completely and solely to Ali Yigit Bireroglu. This includes but is not limited to the source code, the multimedia and                    /
//  other asset files.                                                                                                                               /
//  If you were granted this Intellectual Property for personal use, you are obligated to include this copyright text at all times.                  /
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

import 'package:spring_button/spring_button.dart';
import 'package:cupertino_stackview/cupertino_stackview.dart';

GlobalKey<NavigatorState> navigatorState = GlobalKey<NavigatorState>();

List<PageInfo> pageInfos = [
  PageInfo(
    "Stack I",
    "Lorem Ipsum Dolor",
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
  ),
  PageInfo(
    "Stack II",
    "Sed ut Perspiciatis",
    "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit,.",
  ),
  PageInfo(
    "Stack III",
    "Adipisci Velit",
    "Sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem.",
  ),
  PageInfo(
    "Stack IV",
    "Ut Enim ad Minima",
    "Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur?",
  ),
];

void main() {
  cupertinoStackViewController = CupertinoStackViewController(
    navigatorState,
    {
      "Stack I": (BuildContext context, dynamic parameters) {
        return MyPage(
          0,
        );
      },
      "Stack II": (BuildContext context, dynamic parameters) {
        return MyPage(
          1,
        );
      },
      "Stack III": (BuildContext context, dynamic parameters) {
        return MyPage(
          2,
        );
      },
      "Stack IV": (BuildContext context, dynamic parameters) {
        return MyPage(
          3,
        );
      },
    },
  );
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Cupertino Stack View Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: MyPage(0),
      navigatorKey: navigatorState,
    );
  }
}

class MyPage extends StatefulWidget {
  final int index;

  MyPage(
    this.index, {
    Key? key,
  }) : super(key: key);

  @override
  _MyPageState createState() => _MyPageState();
}

class _MyPageState extends State<MyPage> {
  @override
  Widget build(BuildContext context) {
    cupertinoStackViewController.initialise(MediaQuery.of(context).size.height, MediaQuery.of(context).size.width);
    return CupertinoStackView(
      widget.index == 0,
      pageInfos[widget.index].navigation,
      Scaffold(
        backgroundColor: const Color(0xFFF2F2F7),
        appBar: AppBar(
          backgroundColor: const Color(0xFFF9F9F9),
          leading: widget.index == 0
              ? null
              : SpringButton(
                  SpringButtonType.OnlyScale,
                  Icon(
                    Icons.chevron_left,
                    color: Colors.black,
                  ),
                  useCache: true,
                  onTapUp: (_) {
                    cupertinoStackViewController.back();
                  },
                ),
          title: Text(
            pageInfos[widget.index].navigation,
            style: const TextStyle(
              color: Colors.black,
              fontSize: 32,
              fontWeight: FontWeight.bold,
            ),
          ),
        ),
        body: SingleChildScrollView(
          child: Column(
            children: [
              MyPadding(
                Stack(
                  children: [
                    Container(
                      constraints: BoxConstraints.expand(height: 200),
                      decoration: BoxDecoration(
                        borderRadius: const BorderRadius.all(const Radius.circular(20)),
                        boxShadow: [
                          BoxShadow(
                            color: Colors.black,
                            blurRadius: 50.0,
                            spreadRadius: 2.5,
                          ),
                        ],
                      ),
                    ),
                    ClipRRect(
                      borderRadius: const BorderRadius.all(const Radius.circular(20)),
                      child: Container(
                        constraints: BoxConstraints.expand(height: 200),
                        decoration: BoxDecoration(
                          image: DecorationImage(
                            image: AssetImage(
                              "assets/" + widget.index.toString() + ".jpeg",
                            ),
                            fit: BoxFit.cover,
                          ),
                          boxShadow: [
                            BoxShadow(
                              color: Colors.black.withOpacity(0.75),
                              blurRadius: 100.0,
                              spreadRadius: 5.0,
                            ),
                          ],
                        ),
                      ),
                    ),
                  ],
                ),
              ),
              MyPadding(
                RichText(
                  text: TextSpan(
                    text: pageInfos[widget.index].title + "\n\n",
                    style: const TextStyle(
                      color: Colors.black,
                      fontSize: 32,
                      fontWeight: FontWeight.bold,
                    ),
                    children: [
                      TextSpan(
                        text: pageInfos[widget.index].text,
                        style: const TextStyle(
                          color: Colors.black,
                          fontSize: 16,
                        ),
                      )
                    ],
                  ),
                  textAlign: TextAlign.justify,
                ),
              ),
              if (widget.index != pageInfos.length - 1)
                Align(
                  alignment: Alignment(0.0, 0.5),
                  child: MyPadding(
                    SpringButton(
                      SpringButtonType.OnlyScale,
                      Container(
                        height: 60,
                        width: 175,
                        decoration: const BoxDecoration(
                          color: const Color(0xFF1B1D1E),
                          borderRadius: const BorderRadius.all(const Radius.circular(20)),
                        ),
                        child: Center(
                          child: Text(
                            "Open Stack",
                            style: const TextStyle(
                              color: Colors.white,
                              fontSize: 16,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                        ),
                      ),
                      useCache: true,
                      onTapUp: (_) {
                        cupertinoStackViewController.navigate(pageInfos[widget.index + 1].navigation, context, null);
                      },
                    ),
                  ),
                ),
              if (widget.index == pageInfos.length - 1)
                Align(
                  alignment: Alignment(0.0, 0.5),
                  child: MyPadding(
                    SpringButton(
                      SpringButtonType.OnlyScale,
                      Container(
                        height: 60,
                        width: 175,
                        decoration: const BoxDecoration(
                          color: const Color(0xFF1B1D1E),
                          borderRadius: const BorderRadius.all(const Radius.circular(20)),
                        ),
                        child: Center(
                          child: Text(
                            "Go to Home",
                            style: const TextStyle(
                              color: Colors.white,
                              fontSize: 16,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                        ),
                      ),
                      useCache: true,
                      onTapUp: (_) {
                        cupertinoStackViewController.navigate(pageInfos[0].navigation, context, null);
                      },
                    ),
                  ),
                ),
            ],
          ),
        ),
      ),
      Colors.black,
      isDismissible: widget.index != pageInfos.length - 1,
    );
  }
}

class PageInfo {
  final String navigation;
  final String title;
  final String text;

  PageInfo(
    this.navigation,
    this.title,
    this.text,
  );
}

class MyPadding extends StatelessWidget {
  final Widget _child;

  const MyPadding(
    this._child, {
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(32),
      child: _child,
    );
  }
}


cupertino_stackview/lib/cupertino_stackview_controller.dart
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//  © Cosmos Software | Ali Yigit Bireroglu                                                                                                          /
//  All material used in the making of this code, project, program, application, software et cetera (the "Intellectual Property")                    /
//  belongs completely and solely to Ali Yigit Bireroglu. This includes but is not limited to the source code, the multimedia and                    /
//  other asset files.                                                                                                                               /
//  If you were granted this Intellectual Property for personal use, you are obligated to include this copyright text at all times.                  /
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

import 'package:flutter/cupertino.dart';

import 'package:cupertino_stackview/cupertino_stackview.dart';
import 'package:cupertino_stackview/misc.dart';
import 'package:flutter/material.dart';

///The class that is responsible of all Cupertino StackView related logic.
class CupertinoStackViewController {
  ///The [GlobalKey] used for the [navigatorKey] property of your [CupertinoApp] or [MaterialApp].
  final GlobalKey<NavigatorState> _navigatorState;

  ///The map that is to be used when navigating between the [CupertinoStackView]s in the Cupertino StackView system.
  final Map<String, Function(BuildContext context, dynamic parameters)> _builders;

  ///The duration that is to be used for [CupertinoStackView] animations. The default value corresponds the default Flutter value..
  final Duration duration;

  ///See [initialise].
  late double height;

  ///See [initialise].
  double? width;

  final List<_ListPair> _map = [];
  final List<_ListPair> _siblingMap = [];
  final Map<String, BuildContext?> _contexts = Map<String, BuildContext?>();

  String get currentNavigation => _map.first._navigation;
  BuildContext? get currentContext => _contexts[currentNavigation];

  ///Callback for when any [CupertinoStackView] moves.
  final Function(String, CupertinoStackViewStatus)? onMoved;

  ///Callback for when any [CupertinoStackView] is dismissed.
  final Function(String?)? onDismissed;

  CupertinoStackViewController(
    this._navigatorState,
    this._builders, {
    this.duration: const Duration(milliseconds: 250),
    this.onMoved,
    this.onDismissed,
  });

  bool _isMapped(String navigation) {
    return _map.indexWhere((_ListPair navigationStackStatePair) => navigationStackStatePair._navigation == navigation) > -1;
  }

  bool _isSiblingMapped(String navigation) {
    return _siblingMap.indexWhere((_ListPair navigationStackStatePair) => navigationStackStatePair._navigation == navigation) > -1;
  }

  bool _hasSibling(String navigation) {
    return _siblingMap.indexWhere((_ListPair navigationStackStatePair) => navigationStackStatePair._navigation == navigation) > -1;
  }

  void introduce(String navigation, CupertinoStackViewState cupertinoStackViewState) {
    if (_isMapped(navigation)) {
      _map.firstWhere((_ListPair navigationStackStatePair) => navigationStackStatePair._navigation == navigation)._cupertinoStackViewState = cupertinoStackViewState;
    } else {
      _map.insert(0, _ListPair(navigation, cupertinoStackViewState));
    }
  }

  void introduceAsSibling(String navigation, CupertinoStackViewState cupertinoStackViewState) {
    if (_isSiblingMapped(navigation)) {
      _siblingMap.firstWhere((_ListPair navigationStackStatePair) => navigationStackStatePair._navigation == navigation)._cupertinoStackViewState = cupertinoStackViewState;
    } else {
      _siblingMap.insert(0, _ListPair(navigation, cupertinoStackViewState));
    }
  }

  ///[CupertinoStackViewController] has to be initialised as soon as the application starts.
  ///Use this method with the [MediaQuery.of(context).size.height] and [MediaQuery.of(context).size.width] values.
  void initialise(double height, double width) {
    this.height = height;
    this.width = width;
  }

  void _organise() {
    int _index = 0;
    for (int i = 0; i < _map.length; i++) {
      if (_map[i]._cupertinoStackViewState != null) {
        _map[i]._cupertinoStackViewState!.move(_index == 0
            ? CupertinoStackViewStatus.FRONT
            : _index == 1
                ? CupertinoStackViewStatus.BACK
                : CupertinoStackViewStatus.DISAPPEAR);
        if (onMoved != null) {
          onMoved!(
              _map[i]._navigation,
              _index == 0
                  ? CupertinoStackViewStatus.FRONT
                  : _index == 1
                      ? CupertinoStackViewStatus.BACK
                      : CupertinoStackViewStatus.DISAPPEAR);
        }
        if (_hasSibling(_map[i]._navigation)) {
          _organiseSiblings(_map[i]._navigation, _index);
        }
      }
      _index++;
    }
  }

  void _organiseSiblings(String navigation, int index) {
    List<_ListPair> _siblings = _siblingMap.where((_ListPair navigationStackStatePair) => navigationStackStatePair._navigation == navigation).toList();
    for (int i = 0; i < _siblings.length; i++) {
      if (_siblings[i]._cupertinoStackViewState != null) {
        _siblings[i]._cupertinoStackViewState!.move(index == 0
            ? CupertinoStackViewStatus.FRONT
            : index == 1
                ? CupertinoStackViewStatus.BACK
                : CupertinoStackViewStatus.DISAPPEAR);
        if (onMoved != null) {
          onMoved!(
              _siblings[i]._navigation,
              index == 0
                  ? CupertinoStackViewStatus.FRONT
                  : index == 1
                      ? CupertinoStackViewStatus.BACK
                      : CupertinoStackViewStatus.DISAPPEAR);
        }
      }
    }
  }

  ///Use this method to go back to the primary [CupertinoStackView] in the Cupertino StackView system.
  Future backToPrimary() async {
    while (_map.length != 1) {
      _navigatorState.currentState!.pop();
      _contexts.remove(currentNavigation);
      _map.removeAt(0);
      _organise();
      await Future.delayed(duration);
    }
  }

  ///Use this method to go back one [CupertinoStackView] in the Cupertino StackView system.
  Future back({String? navigation, bool isDismissed: false}) async {
    await navigate(_map[1]._navigation, null, null);
    if (isDismissed && onDismissed != null) {
      onDismissed!(navigation);
    }
  }

  ///Use this method to go to a [CupertinoStackView] in the Cupertino StackView system.
  Future navigate(String targetNavigation, BuildContext? context, dynamic parameters) async {
    if (_isMapped(targetNavigation)) {
      while (_map.first._navigation != targetNavigation) {
        _navigatorState.currentState!.pop();
        _contexts.remove(currentNavigation);
        _map.removeAt(0);
        _organise();
        await Future.delayed(duration);
      }
    } else {
      _map.insert(0, _ListPair(targetNavigation, null));
      _contexts[targetNavigation] = context;
      _organise();
      showCupertinoModalPopup(
          context: context!,
          builder: (BuildContext context) {
            return Stack(
              children: [
                Container(
                  constraints: BoxConstraints.expand(),
                  color: Colors.transparent,
                ),
                _builders[targetNavigation]!(context, parameters),
              ],
            );
          });
    }
  }
}

class _ListPair {
  final String _navigation;
  CupertinoStackViewState? _cupertinoStackViewState;

  _ListPair(
    this._navigation,
    this._cupertinoStackViewState,
  );
}


cupertino_stackview/lib/cupertino_stackview.dart
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//  © Cosmos Software | Ali Yigit Bireroglu                                                                                                          /
//  All material used in the making of this code, project, program, application, software et cetera (the "Intellectual Property")                    /
//  belongs completely and solely to Ali Yigit Bireroglu. This includes but is not limited to the source code, the multimedia and                    /
//  other asset files.                                                                                                                               /
//  If you were granted this Intellectual Property for personal use, you are obligated to include this copyright text at all times.                  /
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';

import 'package:cupertino_stackview/misc.dart';

export 'package:cupertino_stackview/cupertino_stackview_controller.dart';
export 'package:cupertino_stackview/misc.dart';

///The widget that is responsible of all Cupertino StackView related UI.
class CupertinoStackView extends StatefulWidget {
  ///Set this value to true if the [_child] is the first page of the Cupertino StackView system. In theory, there should only be one [CupertinoStackView] with [_isPrimary] set to true.
  final bool _isPrimary;
  bool get isPrimary => _isPrimary;

  ///The navigation path that is to be mapped to this [CupertinoStackView] in the Cupertino StackView system. This value has to be unique for each [CupertinoStackView].
  final String _navigation;

  ///The widget that is to be displayed by this [CupertinoStackView].
  final Widget _child;

  ///The color that is to be shown when this [CupertinoStackView] moves behind the Cupertino StackView system.
  final Color _backgroundColor;

  ///The navigation path that is to link this [CupertinoStackView] to another [CupertinoStackView]. Leave null if you don't want a link.
  final String siblingNavigation;

  ///Set this value to false you don't want the [_child] to be dismissible by vertical dragging.
  final bool isDismissible;

  ///The clipping radius that is to be used to clip this [CupertinoStackView] when it moves behind the Cupertino StackView system. The default value corresponds to the default iOS 13 value.
  final Radius radius;

  ///Set this value to true if you don't want this [CupertinoStackView] to be clipped via [radius] while in the front of the Cupertino StackView system.
  final bool ignoreRadiusWhenFront;

  ///Callback for when this [CupertinoStackView] moves.
  final Function(CupertinoStackViewStatus)? onMoved;

  ///Callback for when this [CupertinoStackView] is dismissed.
  final Function? onDismissed;

  const CupertinoStackView(
    this._isPrimary,
    this._navigation,
    this._child,
    this._backgroundColor, {
    Key? key,
    this.siblingNavigation: "",
    this.isDismissible: true,
    this.radius: const Radius.circular(10.0),
    this.ignoreRadiusWhenFront: false,
    this.onMoved,
    this.onDismissed,
  }) : super(key: key);

  @override
  CupertinoStackViewState createState() {
    return CupertinoStackViewState();
  }
}

class CupertinoStackViewState extends State<CupertinoStackView> with SingleTickerProviderStateMixin {
  final GlobalKey _dismissible = GlobalKey();
  late AnimationController _animationController;
  late Animation<double> _scale;
  late Animation<Offset> _firstOffset;
  late Animation<Offset> _secondOffset;
  CupertinoStackViewStatus? _currentStatus;

  double get _radius => widget.radius.x;
  double get _dynamicRadius => _radius * _animationController.value.clamp(0.0, 1.0);
  Radius get _realRadius => widget.isPrimary || widget.ignoreRadiusWhenFront ? Radius.circular(_dynamicRadius) : widget.radius;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      vsync: this,
      lowerBound: 0.0,
      upperBound: 1.0,
      duration: cupertinoStackViewController.duration,
    );
    _scale = Tween(
      begin: 1.0,
      end: cupertinoStackViewScaleFraction,
    ).animate(
      CurvedAnimation(
        parent: _animationController,
        curve: const Interval(
          0.0,
          0.5,
          curve: Curves.decelerate,
        ),
      ),
    );
    _firstOffset = Tween<Offset>(
      begin: widget._isPrimary ? Offset.zero : Offset(0.0, cupertinoStackViewFirstOffsetFraction * cupertinoStackViewController.height),
      end: Offset.zero,
    ).animate(
      CurvedAnimation(
        parent: _animationController,
        curve: const Interval(
          0.0,
          0.5,
          curve: Curves.decelerate,
        ),
      ),
    );
    _secondOffset = Tween<Offset>(
      begin: Offset.zero,
      end: Offset(0.0, cupertinoStackViewSecondOffsetFraction * cupertinoStackViewController.height),
    ).animate(
      CurvedAnimation(
        parent: _animationController,
        curve: const Interval(
          0.5,
          1.0,
          curve: Curves.decelerate,
        ),
      ),
    );
    _currentStatus = CupertinoStackViewStatus.FRONT;
    if (widget.siblingNavigation.isEmpty) {
      cupertinoStackViewController.introduce(widget._navigation, this);
    } else {
      cupertinoStackViewController.introduceAsSibling(widget.siblingNavigation, this);
    }
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void move(CupertinoStackViewStatus targetStatus) {
    switch (_currentStatus) {
      case CupertinoStackViewStatus.FRONT:
        _animationController.value = 0.0;
        break;
      case CupertinoStackViewStatus.BACK:
        _animationController.value = 0.5;
        break;
      case CupertinoStackViewStatus.DISAPPEAR:
        _animationController.value = 1.0;
        break;
    }

    _currentStatus = targetStatus;

    switch (targetStatus) {
      case CupertinoStackViewStatus.FRONT:
        _animationController.animateTo(0.0);
        break;
      case CupertinoStackViewStatus.BACK:
        _animationController.animateTo(0.5);
        break;
      case CupertinoStackViewStatus.DISAPPEAR:
        _animationController.animateTo(1.0);
        break;
    }
    if (widget.onMoved != null) {
      widget.onMoved!(targetStatus);
    }
  }

  Future<bool> _isDismissed(DismissDirection dismissDirection) async {
    cupertinoStackViewController.back(navigation: widget._navigation, isDismissed: true);
    if (widget.onDismissed != null) {
      widget.onDismissed!();
    }
    return false;
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animationController,
      child: widget._isPrimary
          ? AnimatedBuilder(
              animation: _animationController,
              child: widget._child,
              builder: (BuildContext buildContext, Widget? cachedWidget) {
                return _Clipper(
                  cachedWidget,
                  widget._backgroundColor,
                  _realRadius,
                );
              },
            )
          : widget.isDismissible
              ? Dismissible(
                  key: _dismissible,
                  direction: DismissDirection.down,
                  dismissThresholds: {DismissDirection.down: 0.5},
                  confirmDismiss: _isDismissed,
                  child: _Positioner(
                    widget.ignoreRadiusWhenFront
                        ? AnimatedBuilder(
                            animation: _animationController,
                            child: widget._child,
                            builder: (BuildContext buildContext, Widget? cachedWidget) {
                              return _Clipper(
                                cachedWidget,
                                widget._backgroundColor,
                                _realRadius,
                              );
                            },
                          )
                        : _Clipper(
                            widget._child,
                            widget._backgroundColor,
                            widget.radius,
                          ),
                  ),
                )
              : _Positioner(
                  widget.ignoreRadiusWhenFront
                      ? AnimatedBuilder(
                          animation: _animationController,
                          child: widget._child,
                          builder: (BuildContext buildContext, Widget? cachedWidget) {
                            return _Clipper(
                              cachedWidget,
                              widget._backgroundColor,
                              _realRadius,
                            );
                          },
                        )
                      : _Clipper(
                          widget._child,
                          widget._backgroundColor,
                          widget.radius,
                        ),
                ),
      builder: (BuildContext context, Widget? cachedWidget) {
        return Transform.translate(
          offset: _animationController.value < 0.5 ? _firstOffset.value : _secondOffset.value,
          child: Transform.scale(
            scale: _scale.value,
            alignment: const Alignment(0.0, 1.0),
            child: cachedWidget,
          ),
        );
      },
    );
  }
}

class _Clipper extends StatelessWidget {
  final Widget? _child;
  final Color _backgroundColor;
  final Radius _radius;

  const _Clipper(
    this._child,
    this._backgroundColor,
    this._radius,
  );

  @override
  Widget build(BuildContext context) {
    return ClipRRect(
      borderRadius: BorderRadius.only(
        topLeft: _radius,
        topRight: _radius,
      ),
      child: Container(
        color: _backgroundColor,
        child: _child,
      ),
    );
  }
}

class _Positioner extends StatelessWidget {
  final Widget _child;

  const _Positioner(
    this._child,
  );

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Positioned(
          top: 0.0,
          left: 0.0,
          width: cupertinoStackViewController.width,
          height: cupertinoStackViewSecondHeightFraction * cupertinoStackViewController.height,
          child: _child,
        )
      ],
    );
  }
}
